pip/pip-439.md

```diff
@@ -0,0 +1,410 @@
+          }
+      } else if (schema != null) {
+          typedMessageBuilder = producer.newMessage(schema);
+      } else {
+          typedMessageBuilder = producer.newMessage();
+      }
+    
+      messageBuilder.setUnderlyingBuilder(typedMessageBuilder);
+      return messageBuilder;
+    }
+}
+```
+
+## Asynchronous Functions Support
+
+It's important to note that Pulsar Functions supports asynchronous processing, where functions can return `CompletableFuture` objects. This proposal ensures that transaction support works seamlessly with both synchronous and asynchronous functions.
+
+For asynchronous functions:
+1. The transaction is created at the beginning of message processing, just like for synchronous functions
+2. When the function returns a `CompletableFuture`, the transaction is maintained until the future completes
+ - Any Context-related operations inside of the returned 'CompletableFuture' objects are tied to the correct transaction
+3. When the future completes successfully, the transaction is committed
+4. If the future completes exceptionally, the transaction is aborted
+
+## Batch Processing of Transactions
+
+To optimize performance and reduce the overhead on the Transaction Coordinator, this proposal introduces transaction batching.
+Transaction batching allows multiple incoming messages to be processed within the same transaction, reducing the total number of
+transactions created.
+
+## Transaction Batching Concept
+
+Transaction batching is distinct from Pulsar's message batching. While message batching combines multiple messages into a single "batch
+message" for efficient network transfer, transaction batching processes multiple incoming messages (or batch messages) within the scope
+of a single transaction.
+
+Key benefits of transaction batching include:
+1. **Reduced Load on Transaction Coordinator**: Fewer transactions means less coordination overhead
+2. **Improved Throughput**: Higher message processing capacity with lower per-message overhead
+3. **Optimized Resource Usage**: Better utilization of transaction resources
+4. **Consistent Performance at Scale**: Maintains performance characteristics under high load
+
+### Transaction Batching Parameters
+
+Transaction batching is controlled by two main parameters:
+
+1. **`transactionBatchingMaxEntries`**: The maximum number of entries (incoming messages or batch messages) to process within a single
+transaction before committing it and starting a new one.
+ - An "entry" refers to an incoming batch message which could itself contain multiple individual messages
+ - Default: 1 (recommended minimum to ensure batch index acknowledgment state doesn't span transactions)
+ - Setting this to higher values increases throughput but may impact latency
+
+2. **`transactionBatchingQuietPeriodMs`**: The maximum amount of time to wait for additional messages before committing a transaction if
+`transactionBatchingMaxEntries` is not reached.
+ - Default: 100ms
```

Comment:
```suggestion
 - Default: 1ms
```
It's better to have a shorter default. For Pulsar Client producer, the default `batchingMaxPublishDelay` value is 1ms so that's why I'd go with 1ms.